### PR TITLE
add recipe for tiled

### PR DESCRIPTION
**Pull request by gwbischof**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x]  GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by danielballan** [comment]
I would like to be a maintainer on this feedstock.

**Comment by mrakitin** [comment]
Looks like we don't have all requirements that would correspond to `pip install tiled[all]`, as `grayskull` does not support it. We'll need to manually add all of the requirements from:
- https://github.com/bluesky/tiled/blob/main/requirements-array.txt
- https://github.com/bluesky/tiled/blob/main/requirements-client.txt
- https://github.com/bluesky/tiled/blob/main/requirements-compression.txt
- https://github.com/bluesky/tiled/blob/main/requirements-dataframe.txt
- https://github.com/bluesky/tiled/blob/main/requirements-formats.txt
- https://github.com/bluesky/tiled/blob/main/requirements-server.txt
- https://github.com/bluesky/tiled/blob/main/requirements-xarray.txt

File: recipes/tiled/meta.yaml
Comment (mrakitin) [inline recipes/tiled/meta.yaml:30]:
```suggestion
    - cached-property  # [py<38]
```

Comment (mrakitin) [reply recipes/tiled/meta.yaml:30]: I don't see this suggestion was committed, that's why we are still having trouble with the syntax. Please accept the suggestion, or implement it on your side.

File: recipes/tiled/meta.yaml
Comment (mrakitin) [inline recipes/tiled/meta.yaml:30]: One more attempt 🙂 
```suggestion
    - cached-property  # [py<38]
```

**Comment by mrakitin** [comment]
@conda-forge/staged-recipes, I agree to be listed as a maintainer for this feedstock/recipe.

We had a weird issue in [our CI runs](https://dev.azure.com/nsls2forge/nsls2forge/_build/results?buildId=10447&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=06d62fd2-ab49-5add-9ee5-83b6882d7d2b) for Windows:
```
Executing transaction: ...working... done
ERROR conda.core.link:_execute(730): An error occurred while installing package 'conda-forge::libzlib-1.2.11-h8ffe710_1014'.
Rolling back transaction: ...working... done
Traceback (most recent call last):
  File "D:\a\1\s\.ci_support\build_all.py", line 205, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "D:\a\1\s\.ci_support\build_all.py", line 105, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "D:\a\1\s\.ci_support\build_all.py", line 164, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "C:\Miniconda\lib\site-packages\conda_build\api.py", line 186, in build
    return build_tree(
  File "C:\Miniconda\lib\site-packages\conda_build\build.py", line 3099, in build_tree
    test(pkg, config=metadata.config.copy(), stats=stats)
  File "C:\Miniconda\lib\site-packages\conda_build\build.py", line 2889, in test
    environ.create_env(metadata.config.test_prefix, actions, config=metadata.config,
  File "C:\Miniconda\lib\site-packages\conda_build\environ.py", line 910, in create_env
    execute_actions(actions, index)
  File "C:\Miniconda\lib\site-packages\conda\common\io.py", line 88, in decorated
    return f(*args, **kwds)
  File "C:\Miniconda\lib\site-packages\conda\plan.py", line 321, in execute_actions
    execute_instructions(plan, index, verbose)
  File "C:\Miniconda\lib\site-packages\conda\plan.py", line 533, in execute_instructions
    cmd(state, arg)
  File "C:\Miniconda\lib\site-packages\conda\instructions.py", line 73, in UNLINKLINKTRANSACTION_CMD
    unlink_link_transaction.execute()
  File "C:\Miniconda\lib\site-packages\conda\core\link.py", line 281, in execute
    self._execute(tuple(concat(interleave(itervalues(self.prefix_action_groups)))))
  File "C:\Miniconda\lib\site-packages\conda\core\link.py", line 744, in _execute
    raise CondaMultiError(tuple(concatv(
conda.CondaMultiError: [Errno 13] Permission denied: 'C:\\bld\\tiled-suite_1649349635598\\_test_env\\zlib.dll'
()
```

Is it a known issue, or do you know a workaround?

**Comment by mrakitin** [comment]
@conda-forge/staged-recipes, looks like we've figured out the issue posted above by removing source code packaging for some auxiliary subpackages. This is ready for your review! Thanks.

**Comment by gwbischof** [comment]
@conda-forge-admin, could you review this PR please?

**Comment by gwbischof** [comment]
@conda-forge-admin, please ping team

**Comment by mrakitin** [comment]
@conda-forge/staged-recipes, can it be reviewed/merged soon, please? We need it badly for the deployments at NSLS-II. Thanks!

**Comment by leofang** [comment]
> We need it badly for the deployments at NSLS-II. Thanks!

Ping @carterbox: NSLS-II needs help 🙂

**Comment by mrakitin** [comment]
Thanks for the ping, @leofang 🙂 

Maybe @ocefpaf or @dopplershift will have a chance to review it for us? Any help is appreciated!

File: recipes/tiled/meta.yaml
Comment (carterbox) [inline recipes/tiled/meta.yaml:106]: I assume this is because you are using cached_property from functools in python >=3.8. Is there any other reason besides this that your package cannot be noarch? Would you consider making your conda package noarch by setting the minimum python version to 3.8 or do you still use python 3.7 at NSLS2?

**Review by carterbox — CHANGES_REQUESTED** [review]
I've suggested some changes that would make this package [`noarch`](https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-builds).

File: recipes/tiled/meta.yaml
Comment (gwbischof) [reply recipes/tiled/meta.yaml:106]: This is the only reason that I am aware of.  We still need python 3.7

File: recipes/tiled/meta.yaml
Comment (gwbischof) [reply recipes/tiled/meta.yaml]: @carterbox Do you know why {{ PYTHON }} isn't working and is giving an undefined variable error?

**Review by mrakitin — COMMENTED** [review]
Oops, I realized I did not post my comments to @carterbox's review. Here we go.

File: recipes/tiled/meta.yaml
Comment (carterbox) [reply recipes/tiled/meta.yaml:104]:
```suggestion
        # TODO: Can be noarch when Python 3.7 is EOL in June 2023
        - cached-property  # [py<38]
```
That's fine, I suggest a note like above to remind future maintainers.

File: recipes/tiled/meta.yaml
Comment (carterbox) [reply recipes/tiled/meta.yaml]: My mistake, as @mrakitin mentioned it seems that when the recipe is in multiple-outputs mode, the script parameter is interpreted as a script not as a string which is a script. [The docs](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#outputs-section) indicate that this script parameter is intended for copying files, not building packages.

**Review by carterbox — CHANGES_REQUESTED** [review]
This brings us to the clobber warnings. The Python source files included in tiled-client, tiled-minimal-client, tile-server, tiled-minimal-server should be mutually elusive, but they aren't. This results in clobber warnings when they are all installed multiple times for the tiled complete package and could result in unexpected behavior.

What we need is something like [the matplotlib-feedstock](https://github.com/conda-forge/matplotlib-feedstock/blob/main/recipe/meta.yaml).

1. Add a tiled-base output. This is the *only* output that calls the build script. It provides the python modules, the entry point, and probably only depends on python/click maybe some other base dependencies.
2. The other outputs should all be metapackages that depend on tiled-base and layer additional dependencies on top of it.
3. Ensure that the benchmark module is not installed with tiled because it is outside the tiled namespace.

**Comment by mrakitin** [comment]
Thanks for the useful suggestions, @carterbox! We'll work with @gwbischof on the proposed approach.

Indeed, we used the matplotlib feedstock as an example for this recipe, and it does make sense to package the source code only once to avoid clobbering.

**Comment by gwbischof** [comment]
@carterbox 
I made the changes that you recommended, but we are getting this error on windows:
```
conda.CondaMultiError: [Errno 13] Permission denied: 'C:\\bld\\tiled-suite_1650558569759\\_h_env\\Library\\bin\\libbz2.dll'
```
Do you know how we can fix this?
Thanks!

**Comment by carterbox** [comment]
My Googling says that it might be because conda has run out of space to store all of the environments that have been created (host/build/test) in order to build all of your packages. I suggest removing the tiled-minimal-base package because it is not an actual workload, and it is just a metapackage unlike tiled-base.

**Comment by carterbox** [comment]
Also, consider putting this build into the conda-forge pre-release channel. https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=development%20channel#pre-release-builds

File: recipes/tiled/meta.yaml
Comment (carterbox) [inline recipes/tiled/meta.yaml:257]:
```suggestion
        - pytest  # [not win]
        # broken build of httpcore package fails pipcheck
        - httpcore !=0.13.3
```

**Comment by carterbox** [comment]
If you have docker, you can do linux builds locally so you don't have to wait for the CI.

**Comment by carterbox** [comment]
@gwbischof, I refactored so the build doesn't run out of memory, but I think the external dependencies are now out of date / were not properly specified because the unit tests are failing on unix with errors from tifffile and locket.

**Comment by gwbischof** [comment]
@carterbox could you review/merge this one please?

**Review by carterbox — APPROVED** [review]
Once you have a stable release, you will need to put all future pre-releases into a dev or rc channel.

**Pull request merged**
